myocamlbuild.ml

```diff
@@ -1,44 +1,67 @@
+    At the moment it doesn't, this is unused.
+*)
+let use s = Printf.sprintf "use_obj_%s" s
+
+(** We tag the final binary with dependencies toward the external ocamlfind library. *)
+let pkg s = Printf.sprintf "package(%s)" s
+
+(** Check whether a given flag is set to true in _oasis file
+
+    {!BaseEnvLight} allows use to look inside setup.data for
+    configuration information.
+*)
 let oasis_flag env flag = BaseEnvLight.var_get flag env = "true"
 
+(** The path to the main file, without extension. *)
+let main = "links"
+
+(** For a flag "foo", the internal library is of the form "links-foo". *)
+let lib_of_flag flag = "links-"^flag
+
+(** When a flag is enabled, we tag the main file with a dependency towards
+    the internal corresponding library.
+*)
+let link_conditionally env (flag, file) =
+  let _lib = lib_of_flag flag in
+  if oasis_flag env flag then begin
+    (* We need conditional linking, so only the binaries need the
+       conditional link, not the source file.
+    *)
+    tag_file (main-.-"native") [pkg flag ];
+    tag_file (main-.-"byte")   [pkg flag ];
+
+    (** We plug in dependencies manually. See comment on {!use}.
+        We tag the cm[xo] in order to force the link order.
+    *)
+    dep ["file:"^main-.-"cmx"] [file-.-"cmx"] ;
+    dep ["file:"^main-.-"cmo"] [file-.-"cmo"] ;
+  end
+
+(** The list of flag that have associated libraries
+    that we want conditionally linked. *)
+let flags = [
+  "postgresql", "lib/postgres/pg_database" ;
+  "mysql", "lib/mysql/mysql_database" ;
+  "sqlite", "lib/lite/lite_database" ;
+  "sqlite3", "lib/lite3/lite3_database" ;
+  "monetdb", "lib/monetdb/m5_database" ;
+]
+
 let () = dispatch (fun hook ->
 
-  (* If libraries required by database backends are not installed then
-     ocamlbuild will display a bogus warning that a tag relating to that library
-     is unused.  This silences the warnings. *)
-  let _ = List.map mark_tag_used [ "pkg_postgresql"; "pkg_mysql"; "pkg_sqlite";
-                                   "pkg_sqlite3"; "pkg_mapi" ] in
-
-  (* Run the normal build process *)
-  dispatch_default hook;
-
-  let env_filename = Pathname.basename BaseEnvLight.default_filename in
-  (* JSTOLAREK: I think I can just skip named arguments *)
-  let env = BaseEnvLight.load ~filename:env_filename ~allow_empty:true () in ()
-
-(*
-  if oasis_flag env "postgres" then begin
-      print_string "postgres"
-    end
-  else if oasis_flag env "mysql" then begin
-      print_string "mysql"
-    end;
-  print_string "foo"
-
-  match hook with
-  | After_rules ->
-     if oasis_flag env "is_native" then
-       Command.execute (mv "links.native" "links")
-     else
-       Command.execute (mv "links.byte" "links")
-  | _ -> ()
- *)
- )
+    (* Run the normal build process *)
+    dispatch_default hook ;
+
+    let env_filename = Pathname.basename BaseEnvLight.default_filename in
+    let env = BaseEnvLight.load ~filename:env_filename ~allow_empty:true () in
```

Comment:
I remember a subtlety about that, but I don't remember which exactly. You can try. :)